src/api/classify_element.rs

```diff
@@ -0,0 +1,349 @@
+    /// an until-now full channel) to be awoken, wake them. Return the Async::Ready(Option(Packet))
+    /// 
+    /// #2 Ok(None): this means that the consumer is in tear-down, and we
+    /// will no longer be receivig packets. Return Async::Ready(None) to forward propagate teardown
+    /// 
+    /// #3 Err(TryRecvError::Empty): Packet queue is empty, await the consumer to awaken us with more
+    /// work, and return Async::NotReady to signal to runtime to sleep this task.
+    /// 
+    /// #4 Err(TryRecvError::Disconnected): Consumer is in teardown and has dropped its side of the
+    /// from_consumer channel; we will no longer receive packets. Return Async::Ready(None) to forward
+    /// propagate teardown.
+    /// ###
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.from_consumer.try_recv() {
+            Ok(Some(packet)) => {
+                if let Ok(task) = self.wake_consumer.try_recv() {
+                        task.notify();
+                }
+                Ok(Async::Ready(Some(packet)))
+            },
+            Ok(None) => {
+                Ok(Async::Ready(None))
+            },
+            Err(TryRecvError::Empty) => {
+                let task = task::current();
+                if let Err(_) = self.await_consumer.try_send(task) {
+                    task::current().notify();
+                }
+                Ok(Async::NotReady)
+            },
+            Err(TryRecvError::Disconnected) => {
+                Ok(Async::Ready(None))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::api::element::{Element};
+    use crate::utils::test::packet_generators::{immediate_stream, PacketIntervalGenerator};
+    use crate::utils::test::packet_collectors::ExhaustiveCollector;
+    use crossbeam::crossbeam_channel;
+    use core::time;
+
+    use futures::future::lazy;
+
+    #[allow(dead_code)]
+    struct IdentityElement {
+        id: i32
+    }
+
+    impl Element for IdentityElement {
+        type Input = i32;
+        type Output = i32;
+
+        fn process(&mut self, packet: Self::Input) -> Self::Output {
+            packet
+        }
+    }
+
+    #[allow(dead_code)]
+    struct ClassifyEvenOddElement {
+        id: i32
+    }
+
+    impl ClassifyElement for ClassifyEvenOddElement {
+        type Packet = i32;
+
+        fn classify(&mut self, packet: Self::Packet) -> (usize, Self::Packet) {
+            if (packet % 2) == 0 {
+                (0, packet)
+            } else {
+                (1, packet)
+            }
+        }
+    }   
+
+    #[test]
+    fn one_classify_element() {
+        let default_channel_size = 10;
+        let number_branches = 2;
+        let packets = vec![0, 1, 2, 420, 1337, 3, 4, 5, 6, 7, 8, 9, 11];
+        let packet_generator = PacketIntervalGenerator::new(time::Duration::from_millis(100), packets.clone().into_iter());
```

Comment:
Again for another PR, but I think we should prefer using `immediate_stream` for packet generators to keep tests speedy, and include a test for each component that confirms it can handle pauses in the input stream.

Reply:
:+1:

Reply:
https://github.com/route-rs/route-rs/issues/31